zcash_client_backend/src/data_api.rs

```diff
@@ -20,29 +29,103 @@ use zcash_primitives::{
         components::amount::{Amount, BalanceError, NonNegativeAmount},
         Transaction, TxId,
     },
-    zip32::{AccountId, Scope},
-};
-
-use crate::{
-    address::UnifiedAddress,
-    decrypt::DecryptedOutput,
-    keys::{UnifiedAddressRequest, UnifiedFullViewingKey, UnifiedSpendingKey},
-    proto::service::TreeState,
-    wallet::{Note, NoteId, ReceivedNote, Recipient, WalletTransparentOutput, WalletTx},
-    ShieldedProtocol,
+    zip32::Scope,
 };
+use zip32::DiversifierIndex;
 
 use self::chain::CommitmentTreeRoot;
 use self::scanning::ScanRange;
 
 #[cfg(feature = "transparent-inputs")]
 use zcash_primitives::transaction::components::OutPoint;
 
+use blake2b_simd::{Params as blake2bParams, OUTBYTES as BLAKE2B_OUTBYTES};
+
 pub mod chain;
 pub mod error;
 pub mod scanning;
 pub mod wallet;
 
+/// A hash of a seed used for an HD account.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct HDSeedFingerprint([u8; BLAKE2B_OUTBYTES]);
+
+impl HDSeedFingerprint {
+    /// Generates the fingerprint from a given seed.
+    pub fn from_seed(seed: &SecretVec<u8>) -> Self {
+        const PERSONALIZATION: &[u8] = b"Zcash_HD_Seed_FP";
+        let hash = blake2bParams::new()
+            .personal(PERSONALIZATION)
+            .to_state()
+            .update(seed.expose_secret())
+            .finalize();
+        Self(hash.as_array().to_owned())
+    }
+
+    /// Instantiates the fingerprint from a buffer containing a previously computed fingerprint.
+    pub fn from_bytes(hash: [u8; BLAKE2B_OUTBYTES]) -> Self {
+        Self(hash)
+    }
+
+    /// Returns the bytes of the fingerprint.
+    pub fn as_bytes(&self) -> &[u8; BLAKE2B_OUTBYTES] {
+        &self.0
+    }
+}
+
+/// Represents the identifier for an account that was derived from a ZIP-32 HD seed and account index.
+#[derive(Debug, Clone)]
+pub struct HDSeedAccount(
+    pub HDSeedFingerprint,
+    pub zip32::AccountId,
+    pub UnifiedFullViewingKey,
+);
+
+/// Represents an arbitrary account for which the seed and ZIP-32 account ID are not known
+/// and may not have been involved in creating this account.
+#[derive(Debug, Clone)]
+pub enum ImportedAccount {
+    /// An account that was imported via its full viewing key.
+    Full(UnifiedFullViewingKey),
+}
+
+/// The inputs for adding an account to a wallet.
+#[derive(Debug, Clone)]
+pub enum AccountParameters {
```

Comment:
Nit: I'd prefer this be called `AccountId` - I realize that this is an overload, but I generally interpret `Parameters` to be information that is used for configuration, not identification. Also, these types can go into `zcash_client_backend::wallet`.

Reply:
Oh, actually, on further investigation I would just call this `Account`.

Reply:
Also, requires a `CHANGELOG` entry.